File: LeetCode/Premium/1279. Traffic Light Controlled Intersection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LeetCode.Premium
{
    /// <summary>
    /// out of memory
    /// </summary>
    public class TrafficLight
    {
        private object SyncRoot = new object();
        private int roadIdNum;
        public TrafficLight()
        {
            roadIdNum = 1;
        }

        public void CarArrived(
            int carId,         // ID of the car
            int roadId,        // ID of the road the car travels on. Can be 1 (road A) or 2 (road B)
            int direction,     // Direction of the car
            Action turnGreen,  // Use turnGreen() to turn light to green on current road
            Action crossCar    // Use crossCar() to make car cross the intersection
        )
        {
            lock (SyncRoot)
            {

                if (roadId == roadIdNum)
                {
                    crossCar();
                }
                else
                {
                    roadIdNum = roadId;
                    turnGreen();
                    crossCar();

                }
            }
        }
    }
}

File: LeetCode/Z/1279. Traffic Light Controlled Intersection.java
﻿class TrafficLight
{ 
    private Integer roadIdNum;
   

     public TrafficLight() {
       roadIdNum = 1;
    }

    public void carArrived(
        int carId,           // ID of the car
        int roadId,          // ID of the road the car travels on. Can be 1 (road A) or 2 (road B)
        int direction,       // Direction of the car
        Runnable turnGreen,  // Use turnGreen.run() to turn light to green on current road
        Runnable crossCar    // Use crossCar.run() to make car cross the intersection 
    )
    {
        synchronized (roadIdNum)
        {
            if (roadId != roadIdNum)
            {
               
                turnGreen.run();   
                 roadIdNum = roadId;
            }
			  crossCar.run();
        }
    }
}